[SLP][REVEC] Ignore UserTreeIndex if it is empty.

HanKuanChen opened the pull request:
Previously, the all_of check did not consider the case where the
TreeEntry is empty (i.e., when it is the first entry).

HanKuanChen committed:
[SLP][REVEC] Ignore UserTreeIndex if it is empty.

Previously, the all_of check did not consider the case where the
TreeEntry is empty (i.e., when it is the first entry).

HanKuanChen committed:
[SLP][REVEC] Pre-commit test.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -6449,9 +6449,10 @@ void BoUpSLP::reorderTopToBottom() {
+            // patterns). Only do reorderNodeWithReuses if the user is not
+            // ShuffleVectorInst.
+            if (TE->UserTreeIndex &&
+                isa<ShuffleVectorInst>(TE->UserTreeIndex.UserTE->getMainOp()))
```
Some of the user may not have state, need to add a check for `hasState()` before asking for getMainOp

alexey-bataev replied:
Some entries might have Gather parents

HanKuanChen committed:
remove assert

Previously a TreeEntry may have multiple users. If only one of them is
ShuffleVectorInst, then we don't know how to do. But right now a
TreeEntry can only have 1 user.

HanKuanChen committed:
apply comment

ilovepi commented:
We're seeing what appears to be an MSan test failure after this patch on Linux Aarch64 bots. The patch seems pretty benign to me, but this is the only change in the blame list. Any idea about why that could be?

CC: @vitalybuka in case he knows of something else that could be happening.

Bot link: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-arm64/b8719911097381318689/infra
